WordPress/wordpress-develop
# Improve caching for `theme.json` processing

Related:

- Gutenberg report https://github.com/WordPress/gutenberg/issues/44772
- Trac ticket https://core.trac.wordpress.org/ticket/56467

Props @ockham @aristath

## What

This PR tries to address some performance regressions introduced in WordPress 6.1.

During the beta cycle, we removed some of the existing caching for theme.json processing (see https://github.com/WordPress/wordpress-develop/pull/3408 and https://github.com/WordPress/wordpress-develop/pull/3359). The rationale for this was that this code was now used before all the blocks are registered (aka get template data, etc.) and resulted in stale cache that created issues (see https://github.com/WordPress/gutenberg/issues/44434 and https://github.com/WordPress/gutenberg/issues/44619). We limited it to only reads from the file system. As pointed out in the Gutenberg's issue, this introduced a big impact in performance that this PR is trying to reduce.

## Why

It's a very noticeable degradation that stems from changes done in this cycle.

## How

Caches the calculated data for core, theme, and user based on the blocks that are registered. If the blocks haven't changed since the last time we calculated the data for each origin, we return the cached data. Otherwise, we recalculate it.

Essentially, this brings back the cache we had, but refreshing it when the blocks have changed.

## Performance improvements

The test I'm doing for profiling performance is loading the "Hello World" post as a logged-out user. These are the numbers in the different setups.

| Method                                                     | WordPress 6.0     | WordPress trunk (rev 54449) | This PR          |
| ---------------------------------------------------------- | ----------------- | --------------------------- | ---------------- |
| total time                                                 | 760ms             | 1336ms                      | 787ms            |
| `wp_get_global_settings`                                   | 0.2ms (24 calls)  | 1.2ms (72 calls)            | 0.9ms (72 calls) |
| `wp_get_global_stylesheet`                                 | 1ms (1 call)      | 0.09ms (1 call)             | 0.03ms (1 call)  |
| `wp_get_global_styles_svg_filters`                         | 0.05ms (1 call)   | 0.06ms (1 call)             | 0.04ms (1 call)  |
| `wp_add_global_styles_for_blocks`                          | -                 | 1.2ms (1 call)              | 0.09ms (1 call)  |
| `WP_Theme_JSON_Resolver::get_core_data`                    | 0.3ms (27 calls)  | 53.9ms (76 calls)           | 2.7ms (76 calls) |
| `WP_Theme_JSON_Resolver::get_block_data`                   | -                 | 20ms (76 calls)             | 1.2ms (76 calls) |
| `WP_Theme_JSON_Resolver::get_theme_data`                   | 17ms (49 calls)   | 264ms (98 calls)            | 74ms (98 calls)  |
| `WP_Theme_JSON_Resolver::get_user_data`                    | 0.08ms (27 calls) | 12.8ms (76 calls)           | 0.9ms (76 calls) |
| `WP_Theme_JSON_Resolver::get_merged_data`                  | 12ms (27 calls)   | 121ms (76 calls)            | 108ms (76 calls) |

These numbers can be seen in the cachegrind data that follows:

- WordPress 6.0: [cachegrind.out.6-0.gz](https://github.com/WordPress/wordpress-develop/files/9749720/cachegrind.out.6-0.gz)
- Trunk at rev 54449: [cachegrind.out.6-1-trunk-54449.gz](https://github.com/WordPress/wordpress-develop/files/9749719/cachegrind.out.6-1-trunk-54449.gz)
- This PR: [cachegrind.out.6-1-pr-3418.gz](https://github.com/WordPress/wordpress-develop/files/9749722/cachegrind.out.6-1-pr-3418.gz)

You can take this data and open it with any visualizer of your choice (kcachegrind, webgrind, etc.). As with the client performance numbers, the exact ms are not as important as the variance between them. I also share below how to reproduce this for anyone interested.

## How to reproduce the performance results

Enable xdebug profiling:

- Clone the https://github.com/WordPress/wordpress-develop repo. Apply this patch to the configuration:
- Configure the environment so xdebug profiles the application by URL:
  - Edit the `.env` file and set `LOCAL_PHP_XDEBUG=true`, and `LOCAL_PHP_XDEBUG_MODE=develop,debug,profile`. For WordPress 6.0, copy `XDEBUG_MODE=develop,debug,profile` in the `docker-compose.yml` file instead (see below).
  - Edit the `docker-compose.yml` file and add `XDEBUG_CONFIG=profiler_output_name=cachegrind.out.%R`.
- Open the browser and visit `http://localhost:8889/?p=1` (this loads the "Hello World" post).
- As a result of this configuration, the docker container should have created a few files named `/tmp/cachegrind.out.*`. Verify those files exist. You may use this command <pre> docker exec -it \`docker ps -f name=wordpress-develop_php -q\` ls -l /tmp</pre>

Read the profiling reports using webgrind:

- Download or clone https://github.com/jokkedk/webgrind in `wp-content/plugins/webgrind` of the `wordpress-develop` repo.
- Start the `wordpress-develop` environment normally.
- Visit http://localhost:8889/wp-content/plugins/webgrind/ and check the results for this branch. You'll have to do the same for checking the results in `trunk` and in the 6.0 branch.

Alternatively, you can read the profiling reports using kcachegrind (linux, probably qcachegrind can do the same for mac & windows?), which allows for loading the xdebug files directly so you can easily compare branches:

- Copy the files generated by xdebug for this branch (`/tmp/cachegrind.out.*`) somewhere outside the docker container. Do the same for the files generated for trunk and 6.0 (be aware that the files will overwrite each other, so copy before changing branches).
- For example, I copy the files by doing:
```sh
docker cp `docker ps -f name=wordpress-develop_php -q`:'/tmp/cachegrind.out.__p=1.gz' /tmp/cachegrind.out.<6-0 or trunk or this-branch>.gz
```
- Open the app and check the results.

## How to test

Make sure automated test pass.

Manually test that global styles still works as expected by using the TT2 or TT3 themes and updating some block styles via the global styles sidebar.
